File: LeetCode/Java Solutions/Design Underground System/design.java
import java.util.HashMap;
import java.util.Map;

class UndergroundSystem {
    private Map<Integer, CheckInData> checkIns;
    private Map<String, StationData> stationData;

    public UndergroundSystem() {
        checkIns = new HashMap<>();
        stationData = new HashMap<>();
    }

    public void checkIn(int id, String stationName, int t) {
        checkIns.put(id, new CheckInData(stationName, t));
    }

    public void checkOut(int id, String stationName, int t) {
        CheckInData checkInData = checkIns.get(id);
        String route = checkInData.stationName + "-" + stationName;
        int travelTime = t - checkInData.time;

        stationData.putIfAbsent(route, new StationData());
        StationData station = stationData.get(route);
        station.totalTime += travelTime;
        station.totalTrips++;
    }

    public double getAverageTime(String startStation, String endStation) {
        String route = startStation + "-" + endStation;
        StationData station = stationData.get(route);
        return (double) station.totalTime / station.totalTrips;
    }

    private static class CheckInData {
        String stationName;
        int time;

        public CheckInData(String stationName, int time) {
            this.stationName = stationName;
            this.time = time;
        }
    }

    private static class StationData {
        int totalTime;
        int totalTrips;
    }
}

File: LeetCode/Python Solutions/Design Underground System/design.py
class UndergroundSystem:
    def __init__(self):
        self.checkIns = {}
        self.stationData = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkIns[id] = {"stationName": stationName, "time": t}

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        checkInData = self.checkIns[id]
        route = checkInData["stationName"] + "-" + stationName
        travelTime = t - checkInData["time"]

        if route not in self.stationData:
            self.stationData[route] = [0, 0]
        self.stationData[route][0] += travelTime
        self.stationData[route][1] += 1

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        route = startStation + "-" + endStation
        data = self.stationData[route]
        return data[0] / data[1]


if __name__ == "__main__":
    undergroundSystem = UndergroundSystem()
    undergroundSystem.checkIn(1, "A", 0)
    undergroundSystem.checkOut(1, "B", 5)
    averageTime = undergroundSystem.getAverageTime("A", "B")
    print("Average Time:", averageTime)
